src/main/java/org/jenkinsci/remoting/engine/JnlpProtocol4Handler.java

```diff
@@ -0,0 +1,384 @@
+                JnlpClientDatabase.ValidationResult validation = event.getCertificate() == null
+                        ? JnlpClientDatabase.ValidationResult.UNCHECKED
+                        : clientDatabase.validateCertificate(clientName, event.getCertificate());
+                switch (validation) {
+                    case IDENTITY_PROVED:
+                        // no-op, we trust the certificate as being from the client
+                        break;
+                    case INVALID:
+                        LOGGER.log(Level.WARNING,
+                                "An attempt was made to connect as {0} from {1} with an invalid client certificate",
+                                new Object[]{clientName, event.getSocket().getRemoteSocketAddress()});
+                        throw new ConnectionRefusalException("Authentication failure");
+                    default:
+                        String secretKey = clientDatabase.getSecretOf(clientName);
+                        if (secretKey == null) {
+                            // should never get hear unless there is a race condition in removing an entry from the DB
+                            throw new ConnectionRefusalException("Unknown client name: " + clientName);
+                        }
+                        if (!secretKey.equals(headers.get(JnlpConnectionState.SECRET_KEY))) {
+                            LOGGER.log(Level.WARNING,
+                                    "An attempt was made to connect as {0} from {1} with an incorrect secret",
+                                    new Object[]{clientName, event.getSocket().getRemoteSocketAddress()});
+                            throw new ConnectionRefusalException("Authorization failure");
+                        }
+                        break;
+                }
+            }
+            event.fireAfterProperties(headers);
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void onChannel(@Nonnull final Channel channel) {
+            channel.addListener(this);
+            threadPool.execute(new Runnable() {
+                @Override
+                public void run() {
+                    if (!channel.isClosingOrClosed()) {
+                        event.fireAfterChannel(channel);
+                    }
+                }
+            });
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        @Nonnull
+        public ChannelBuilder decorate(@Nonnull ChannelBuilder builder) {
+            if (!client) {
+                builder.withMode(Channel.Mode.NEGOTIATE);
+            }
+            event.fireBeforeChannel(builder);
+            return event.getChannelBuilder();
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void onClosed(Channel channel, IOException cause) {
+            event.fireChannelClosed(cause);
+            try {
+                event.getSocket().close();
```

Comment:
so we do not remove listener in this method like it's being done for nClosed(ProtocolStack stack, IOException cause)